---
title: "Tutorial_TADkit_R_package"
output:
    html_document: 
      self_contained: true
      number_sections: yes
      toc: yes
editor_options:
  chunk_output_type: inline
vignette: >
  %\VignetteIndexEntry{Turorial_TADkit_R_package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>")
```

# Preamble

The main objective of the TADkit R package is to make the visualization of data associated with HiC analyses accessible to biologists without the need for bioinformatics skills. To this end, two pairs of functions have been created to visualize :

* domains such as TADs (topological Associated Domain) or compartments (compartment A and B),
* interaction matrices.

Other tools has been added in order to analysed genomic annotations in the light of the 3D organisation.

In this tutorial, we are going to start with the visualization of domains and matrices with the datas included in the package. Note that most of the functions refer to the so called TAD but it also work for other kind of domains (e.g compartments). 

# Data format

First, we will see the different types of data and their formats that can be visualized with TADkit. Datas of 2 chromosomes and 2 cows are available within the package.
Let's start by loading the packages needed for that tutorial.
```{r setup, warning=FALSE, message=FALSE}
library(TADkit)
library(dplyr)
library(GenomicRanges)
library(GenomicFeatures)
library(ggplot2)
```

## matrix

Interaction count (HiC matrix) for a genome is a very large size and can be store in a multiple formats. HiC matrix is symmetrical with respect to the diagonal, so only the upper part of the diagonal is loaded in memory for only one chromosome (intra chromosomal interactions).

### data frame

A basic storage is a compressed data frame for each chromosome in which each row and columns represent a bin. Bovine matrices for chromosome 25 and 26 are available in the TADkit package as a sparse `dgCMatrix` format (i.e upper part of the matrix without the zeros). 
```{r}
class(TADkit::matrix_1_chr25_50kb)

#raw count between bins 60 to 70
TADkit::matrix_1_chr25_50kb[60:70,60:70]
```

Let's write the matrix as a gzip data frame on the current directory and load in R as matrix format:
```{r}
write.table(as.data.frame(as.matrix(TADkit::matrix_1_chr25_50kb)), #dgCMatrix 2 data frame
            "./matrix_1_chr25_50kb.df", 
            row.names = TRUE, #add row names
            col.names = TRUE, #add column names
            quote = FALSE, sep = "\t") 
```

Those matrices can be load in R as a matrix:
```{r}
mat1.df = read.table("./matrix_1_chr25_50kb.df",
                 skip = 1, #skip col.names
                 sep = "\t",
                 row.names = 1 #specify the columns number with row names
                 ) 
mat1.mat = as.matrix(mat1.df) # translate to matrix
mat1.dgcmat = as(mat1.mat, "dgCMatrix") #translate to dgCMatrix
```

### .cool / .mcool

One of the most common HiC array storage formats is the .cool or .mcool format which store matrices for 1 resolution and multiple resolutions respectively. To load HiC matrix for one chromosome from cool/.mcool as a matrix (`dgCMatrix`) use `coolFetch()`:
```{r, eval=FALSE}
#read .cool file
my_matrix.10kb = coolFetch("my_matrix.10kb.cool", chr = "chr_name")

#read .mcool file
my_matrix.10kb = coolFetch("my_matrix.mcool", chr = "chr_name", bin.width = 10e3)
```

## .bed

Domains are usely stored in bed file format which contains at least 3 columns (chromosome name, the start and the end of the TAD). 
```{r}
head(TADkit::tad_1_10kb.bed) #TAD for bovine 1 estimated from 10kb matrix
```
Other columns can be added like the stand, any score or character. In most of the functions bed files must be stored in `GRanges` in which we can store the size of the chromosomes we will need.  
To create a `GRanges` with the size of the chromosomes use `dataframe2grange()`:
```{r}
#chromosomes sizes
TADkit::chromsize

#create GRanges with chr sizes:
tad_1_10kb.gr = dataframes2grange(TADkit::tad_1_10kb.bed, TADkit::chromsize)
tad_1_10kb.gr

#chromsize within GRanges object
seqlengths(tad_1_10kb.gr)
```
TADs are often stored in the form of domains and each domain gives the beginning and the end of a TAD. Sometime TADs are stored in the form of boundaries which gives bins in which borders are located. The width of the boundaries corresponds to the size of the bins in the matrix (i.e 10kb in our datas).
```{r}
#TADs as a form of boundaries:
boundaries = data.frame(chr = seqnames(tad_1_10kb.gr),
                      start = start(tad_1_10kb.gr) - 5e3,
                      end = start(tad_1_10kb.gr) + 5e3)
head(boundaries)
```
Then boundaries can be translated as domains with `boundary2domain()`:
```{r}
boundary2domain(boundaries)
```

In addition to TADs, we can also plot any type of annotations. For this tutorial we will use genes annotations available on Ensembl:
```{r}
txdb <- makeTxDbFromBiomart(biomart="ensembl", dataset="btaurus_gene_ensembl")
genomic.gr = genes(txdb, columns=c("TXTYPE"))
genomic.gr
```

## .bedgraph

Bedgraph are used to store a score for each bin, like insulation score. Insulation score (IS) is defined for a bin as an average number of intyeractions that occur across this bin in some vicinity of the bin (Crane et al., 2015). Insulation scores are used to call TAD boundaries.
In TADkit plot functions, bedgraph inputs can be in 3 formats:

* `dataframe`,
* `GRanges`,
* or the path of the file.

To avoid having to load too much data in the R environment, we will see in this tutorial how to use the paths:
```{r}
head(TADkit::IS_1_10kb.bedgraph)

#write bedgraph in a file
write.table(TADkit::IS_1_10kb.bedgraph, "./IS_1_10kb.bedgraph", col.names = FALSE, quote = FALSE, row.names = FALSE, sep = "\t")
```

## .bigwig

Similar to bedgraph, bigwig files are used to store a score for each interval in an indexed and compressed form. 
Insulation scores stored as a bigwig format can be translated to a `GRranges` with `bw2grange()`.

For this tutorial we are going to use coverage data from RNA sequencing experiment available on Ensembl:
```{r}
#download expression data (RNAseq) of bovine:
download.file("http://ftp.ensembl.org/pub/release-104/bamcov/bos_taurus/genebuild/ARS-UCD1.2.ENA.heart.1.bam.bw", destfile = "./rna_seq_bov.bw", method = "curl") #if it fails: remove the method parameter or download it manually in the current directory
```

# Domains plot

Two functions have been built to plot TADs: 
for one indiviual `TADplot()` and multiple individual `mTADplot()`. 

In addition to domains, few others tracks can be plotted:

* bedgraph to plot bin scores (e.g insulation score...),
* bigwig to plot coverage datas (e.g RNAseq...),
* bed to plot annotations (e.g genes...).

## TADplot

Let start with the most basic usage:
```{r, fig.height = 1.5, fig.width = 8}
TADplot(tad.gr = tad_1_10kb.gr, chr = 25, start = 13e6, stop = 15e6)
```

Note that the area is extended to the first and last TAD of the window.

Add insulation score (bedgraph), RNAseq (bigwig) and genomic annotations (genes grouped according to the "TXTYPE" column, i.e first metadata column) to the plot:
```{r, fig.height = 5, fig.width = 8}
TADplot(tad.gr = tad_1_10kb.gr, chr = 25, start = 13e6, stop = 15e6,
        bedgraph = IS_1_10kb.bedgraph,
        bigwig.path = "./rna_seq_bov.bw", 
        annot.gr = genomic.gr, 
        annot.col = 1, #columns to group annotations
        bigwig.yaxis = "log2" #log2 of RNAseq values
        )
```

## mTADplot

`mTADplot()` allow to plot 1 or more individual/sample in a same plot. All tracks inputs must be stored in a list, and all object in the list must have names:

Datas for the 2 bovines:
```{r}
#create GRanges with TADs
tad_2_10kb.gr = dataframes2grange(TADkit::tad_2_10kb.bed, TADkit::chromsize) 

#write bedgraph as file
write.table(TADkit::IS_2_10kb.bedgraph, "./IS_2_10kb.bedgraph", col.names = FALSE, quote = FALSE, row.names = FALSE, sep = "\t") 
write.table(TADkit::PC1_1_50kb.bedgraph, "./PC1_1_50kb.bedgraph", col.names = FALSE, quote = FALSE, row.names = FALSE, sep = "\t") 
write.table(TADkit::PC1_2_50kb.bedgraph, "./PC1_2_50kb.bedgraph", col.names = FALSE, quote = FALSE, row.names = FALSE, sep = "\t") 
```

### Create list

Let's create lists for the 2 individuals (then named as "bov1" and "bov2"):

* TADs:
```{r}
#list of TADs
tad.lst = list(tad_1_10kb.gr, tad_2_10kb.gr)
names(tad.lst) = c("bov1", "bov2")

#insulation score (path) list
IS.lst = list("./IS_1_10kb.bedgraph", "./IS_2_10kb.bedgraph")
names(IS.lst) = c("bov1", "bov2")
```

* insulation scores:
```{r}
#insulation score (path) list
IS.lst = list("./IS_1_10kb.bedgraph", "./IS_2_10kb.bedgraph")
names(IS.lst) = c("bov1", "bov2")

IS.lst = list(makeGRangesFromDataFrame(IS_1_10kb.bedgraph, keep.extra.columns = T), makeGRangesFromDataFrame(IS_2_10kb.bedgraph, keep.extra.columns = T))
names(IS.lst) = c("bov1", "bov2")
```

Do the plot:
```{r, fig.height = 2.5, fig.width = 8}
mTADplot2(tad.lst = tad.lst, chr = 25, start = 13e6, stop = 15e6,
        bedgraph.lst = IS.lst, bedgraph.name = "IS")
```

### Create multiple lists

In case of several tracks of bedgraph, we can create a list containing few other lists.
Let's create 2 lists with the insulation scores and the PC1 scores:
```{r}
#insulation scores (path) as list
IS.lst = list("./IS_1_10kb.bedgraph", "./IS_2_10kb.bedgraph")
names(IS.lst) = c("bov1", "bov2")

#PC1 scores (path) as list
PC1.lst = list("./PC1_1_50kb.bedgraph", "./PC1_2_50kb.bedgraph")
names(PC1.lst) = c("bov1", "bov2")
```

Now let's create a list with those 2 lists:
```{r}
#multiple lists fo bedgraph
bedgraphs.lst = list(IS.lst, PC1.lst)
names(bedgraphs.lst) = c("IS", "PC1")

bedgraphs.lst
```

Do the plot:
```{r, fig.height = 3.5, fig.width = 8}
mTADplot2(tad.lst = tad.lst, chr = 25, start = 13e6, stop = 15e6,
        bedgraph.lst = bedgraphs.lst)
```

## Options

When you master TADplot(), mTADplot() is much more powerful and flexible to make the desired graphics. Moreover there are many options to change some of the graphic parameters:

### bigwigPath.lst	

* bigwig.binsize: Bin sizes for the histogram. Default = 1e3.

* bigwig.xaxis: Function used to transform the x-axis among each bigwig.binsize. Defaults = "median". Alternatively, other predefined functions can be supplied as character ("mean", "median", "sum", "min", "max" or "extreme").

* bigwig.chr: Chromosome name used to filter chromosome names that can be different from chr (e.g chr = "1" and bigwig.chr = "chr1"). Default = NULL to used the same name as chr.

* bigwig.yaxis: Function used to transforming the y-axis values. Default = NULL. Use "log2" to use the function log2(x + 1) to transform the y-axis (for RNA seq) or provide any other function.

### annotation.lst

* annot.col: Column number of the metadata from annot.gr file(s) used to group the annotation tracks. Default = NULL and the name of each annotation is added.

### bedgraphPath.lst	

* bedgraph.name: Name of the bedgraph track when there is only one track (default = "bedgraph"). Otherwise it takes the names of each list.

* bedgraph_outliers: Ratio to remove outliers of all each bedgraph values. Default is 0 (ie no filter). To remove the first and last 2 percentiles (i.e extreme values / outliers) use 0.02.

# Matrix plot

Two functions have been built to plot interaction count (HiC matrix) for one matrix `MATplot()` and 2 matrices `mMATplot()` (on the upper or lower part) as a ggplot graph. 
In addition to matrices, 3 others tracks can be plotted:

* bed file type to visualized domains as lines or a triangles,
* bedpe file type to hightlight interactions between 2 areas (loops).

## MATplot

### matrix + triangles

Plot of the matrix at 50kb resolution with TADs as triangles on the upper part of the matrix.
```{r, fig.height = 7, fig.width = 7}
MATplot(matrix_1_chr25_50kb, start = 10e6, stop = 30e6,
        bin.width = 50e3, log2 = T,
        tad.upper.tri = tad_1_10kb.gr,
        tad.chr = 25, #filter TADs for chr 25
        scale.colors = "H", #color of matrix, try "D" or "H"
        annotations.color = "red")+
  ggtitle("log2(raw count): chr25 bov1")
```

### matrix + triangles + loops + lines

bedpe file are used to highlight 2 areas. For example let's create a bedpe file to highlight a region on chromosome 25 between :

* 11Mb to 13Mb,
* 26Mb to 27Mb:
```{r}
bedpe = data.frame(chr1 = "25", start1 = 11e6, end1 = 13e6,
                   chr2 = "25", start2 = 26e6, end2 = 27e6)
bedpe
```

In addition to bedpe file, let's see 2 others functions:

* `matObsExp()` to produce the observed / expected ratio of interaction count,
* `PC1calling()` to call the compartments A and B from PC1 values.

```{r, fig.height = 7, fig.width = 7}
MATplot(matObsExp(matrix_1_chr25_50kb), start = 10e6, stop = 30e6,
        bin.width = 50e3, log2 = T,
        tad.upper.tri = tad_1_10kb.gr,
        tad.chr = 25, #filter TADs for chr 25
        scale.colors = "OE", 
        annotations.color = "red",
        loop.bedpe = bedpe,
        tad.upper.line = PC1calling(PC1_1_50kb.bedgraph), #compartments
        tad.line.col = 1 #use the fist metadata columns with vectors A and B
        )+
  ggtitle("log2(obs/exp): chr25 bov1")
```

# Clear files

```{r}
file.remove(
  grep(list.files(full.names = TRUE), pattern = "Rmd", invert = TRUE, value = TRUE))
```



# Analysis

In progress...















































